File: board.py
"""
board.py
The Qwirkle board: an unbounded grid of (x, y) -> Tile.

A "move" is a list of (x, y, tile) triples representing everything a
player places down on a single turn. This module validates moves against
the real Qwirkle placement rules and computes the resulting score.
"""

from collections import defaultdict
from tiles import Tile


class InvalidMove(Exception):
    """Raised (as a reason string, not actually thrown during normal play)
    to explain why a candidate move is illegal."""
    pass


class Board:
    def __init__(self):
        self.cells = {}  # (x, y) -> Tile

    def is_empty(self):
        return len(self.cells) == 0

    def get(self, pos):
        return self.cells.get(pos)

    def occupied(self, pos):
        return pos in self.cells

    # ---------- line helpers ----------

    @staticmethod
    def _line_cells(temp_cells, pos, direction):
        """Return the contiguous run of occupied cells through `pos` in
        `direction` ('h' = varying x, same y; 'v' = varying y, same x),
        as a sorted list of (pos, tile)."""
        x, y = pos
        step = (1, 0) if direction == "h" else (0, 1)
        cells = [pos]

        cx, cy = x - step[0], y - step[1]
        while (cx, cy) in temp_cells:
            cells.insert(0, (cx, cy))
            cx -= step[0]
            cy -= step[1]

        cx, cy = x + step[0], y + step[1]
        while (cx, cy) in temp_cells:
            cells.append((cx, cy))
            cx += step[0]
            cy += step[1]

        return [(p, temp_cells[p]) for p in cells]

    @staticmethod
    def _line_valid(line):
        """A line (list of (pos, tile)) is valid if every tile shares a
        color (with all-distinct shapes) or shares a shape (with all
        distinct colors), has no duplicate tile, and is at most 6 long."""
        tiles = [t for _, t in line]
        if len(tiles) > 6:
            return False
        sigs = {(t.color, t.shape) for t in tiles}
        if len(sigs) != len(tiles):
            return False  # duplicate tile in the line
        colors = {t.color for t in tiles}
        shapes = {t.shape for t in tiles}
        return len(colors) == 1 or len(shapes) == 1

    # ---------- validation ----------

    def validate_move(self, move):
        """
        move: list of (x, y, tile)
        Returns (ok: bool, reason_or_lines):
            if ok -> (True, lines) where lines is the list of resulting
                     lines (each a list of (pos, tile)) touched by this move
            if not ok -> (False, "human readable reason")
        """
        if not move:
            return False, "No tiles placed."

        positions = [(x, y) for x, y, _ in move]
        if len(set(positions)) != len(positions):
            return False, "Can't place two tiles on the same square."

        for pos in positions:
            if self.occupied(pos):
                return False, f"Square {pos} is already occupied."

        tile_map = {(x, y): t for x, y, t in move}
        temp_cells = dict(self.cells)
        temp_cells.update(tile_map)

        if len(move) > 1:
            xs = {x for x, y in positions}
            ys = {y for x, y in positions}
            if len(xs) == 1:
                main_dir = "v"
            elif len(ys) == 1:
                main_dir = "h"
            else:
                return False, "All tiles in a turn must be placed in a single straight line."
        else:
            main_dir = None

        lines = []

        if main_dir:
            main_line = self._line_cells(temp_cells, positions[0], main_dir)
            main_line_positions = {p for p, _ in main_line}
            for p in positions:
                if p not in main_line_positions:
                    return False, "Tiles must form one contiguous line (no gaps)."
            if not self._line_valid(main_line):
                return False, "The main line has a color/shape conflict or duplicate."
            lines.append(main_line)

            perp_dir = "h" if main_dir == "v" else "v"
            for p in positions:
                pline = self._line_cells(temp_cells, p, perp_dir)
                if len(pline) > 1:
                    if not self._line_valid(pline):
                        return False, f"Cross line through {p} has a color/shape conflict or duplicate."
                    lines.append(pline)
        else:
            pos = positions[0]
            for d in ("h", "v"):
                line = self._line_cells(temp_cells, pos, d)
                if len(line) > 1:
                    if not self._line_valid(line):
                        return False, "Placement conflicts with an existing line."
                    lines.append(line)

        if not self.is_empty():
            touches_existing = any(
                any(p in self.cells for p, _ in line) for line in lines
            )
            if not touches_existing:
                return False, "Your tiles must connect to tiles already on the board."

        return True, lines

    def score_move(self, lines):
        """Given the validated `lines` from validate_move, compute the score."""
        total = 0
        for line in lines:
            n = len(line)
            if n < 2:
                continue
            total += n
            if n == 6:
                total += 6  # Qwirkle bonus
        return total

    def apply_move(self, move):
        for x, y, t in move:
            self.cells[(x, y)] = t

    # ---------- convenience for bots/GUI ----------

    def legal_anchor_squares(self):
        """All empty squares adjacent to at least one occupied square --
        useful for bots scanning for candidate placements."""
        if self.is_empty():
            return {(0, 0)}
        anchors = set()
        for (x, y) in self.cells:
            for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                p = (x + dx, y + dy)
                if p not in self.cells:
                    anchors.add(p)
        return anchors

    def bounds(self):
        """(min_x, min_y, max_x, max_y) of played tiles, or None if empty."""
        if self.is_empty():
            return None
        xs = [x for x, y in self.cells]
        ys = [y for x, y in self.cells]
        return min(xs), min(ys), max(xs), max(ys)

File: board.js
// board.js
// The Qwirkle board: an unbounded grid of "x,y" -> Tile.
//
// A "move" is an array of [x, y, tile] triples representing everything a
// player places down on a single turn. This module validates moves
// against the real Qwirkle placement rules and computes the resulting
// score. Direct port of board.py.

function posKey(x, y) {
  return `${x},${y}`;
}

function parseKey(key) {
  const [x, y] = key.split(",").map(Number);
  return [x, y];
}

export class Board {
  constructor() {
    this.cells = new Map(); // "x,y" -> Tile
  }

  isEmpty() {
    return this.cells.size === 0;
  }

  get(x, y) {
    return this.cells.get(posKey(x, y)) || null;
  }

  occupied(x, y) {
    return this.cells.has(posKey(x, y));
  }

  // ---------- line helpers ----------

  // Returns the contiguous run of occupied cells through (x, y) in
  // `direction` ('h' = varying x, same y; 'v' = varying y, same x), as a
  // sorted array of {x, y, tile}.
  static _lineCells(tempCells, x, y, direction) {
    const step = direction === "h" ? [1, 0] : [0, 1];
    const cellsList = [[x, y]];

    let cx = x - step[0], cy = y - step[1];
    while (tempCells.has(posKey(cx, cy))) {
      cellsList.unshift([cx, cy]);
      cx -= step[0];
      cy -= step[1];
    }

    cx = x + step[0]; cy = y + step[1];
    while (tempCells.has(posKey(cx, cy))) {
      cellsList.push([cx, cy]);
      cx += step[0];
      cy += step[1];
    }

    return cellsList.map(([px, py]) => ({ x: px, y: py, tile: tempCells.get(posKey(px, py)) }));
  }

  // A line (array of {x,y,tile}) is valid if every tile shares a color
  // (with all-distinct shapes) or shares a shape (with all distinct
  // colors), has no duplicate tile, and is at most 6 long.
  static _lineValid(line) {
    const tiles = line.map((c) => c.tile);
    if (tiles.length > 6) return false;
    const sigs = new Set(tiles.map((t) => t.sig()));
    if (sigs.size !== tiles.length) return false; // duplicate tile in the line
    const colors = new Set(tiles.map((t) => t.color));
    const shapes = new Set(tiles.map((t) => t.shape));
    return colors.size === 1 || shapes.size === 1;
  }

  // ---------- validation ----------

  // move: array of [x, y, tile]
  // Returns { ok: true, lines } or { ok: false, reason }
  // `lines` is the list of resulting lines (each an array of {x,y,tile})
  // touched by this move.
  validateMove(move) {
    if (!move || move.length === 0) {
      return { ok: false, reason: "No tiles placed." };
    }

    const positions = move.map(([x, y]) => [x, y]);
    const posSet = new Set(positions.map(([x, y]) => posKey(x, y)));
    if (posSet.size !== positions.length) {
      return { ok: false, reason: "Can't place two tiles on the same square." };
    }

    for (const [x, y] of positions) {
      if (this.occupied(x, y)) {
        return { ok: false, reason: `Square (${x}, ${y}) is already occupied.` };
      }
    }

    const tempCells = new Map(this.cells);
    for (const [x, y, t] of move) tempCells.set(posKey(x, y), t);

    let mainDir = null;
    if (move.length > 1) {
      const xs = new Set(positions.map(([x]) => x));
      const ys = new Set(positions.map(([, y]) => y));
      if (xs.size === 1) mainDir = "v";
      else if (ys.size === 1) mainDir = "h";
      else return { ok: false, reason: "All tiles in a turn must be placed in a single straight line." };
    }

    const lines = [];

    if (mainDir) {
      const [x0, y0] = positions[0];
      const mainLine = Board._lineCells(tempCells, x0, y0, mainDir);
      const mainLinePositions = new Set(mainLine.map((c) => posKey(c.x, c.y)));
      for (const [x, y] of positions) {
        if (!mainLinePositions.has(posKey(x, y))) {
          return { ok: false, reason: "Tiles must form one contiguous line (no gaps)." };
        }
      }
      if (!Board._lineValid(mainLine)) {
        return { ok: false, reason: "The main line has a color/shape conflict or duplicate." };
      }
      lines.push(mainLine);

      const perpDir = mainDir === "v" ? "h" : "v";
      for (const [x, y] of positions) {
        const pline = Board._lineCells(tempCells, x, y, perpDir);
        if (pline.length > 1) {
          if (!Board._lineValid(pline)) {
            return { ok: false, reason: `Cross line through (${x}, ${y}) has a color/shape conflict or duplicate.` };
          }
          lines.push(pline);
        }
      }
    } else {
      const [x, y] = positions[0];
      for (const d of ["h", "v"]) {
        const line = Board._lineCells(tempCells, x, y, d);
        if (line.length > 1) {
          if (!Board._lineValid(line)) {
            return { ok: false, reason: "Placement conflicts with an existing line." };
          }
          lines.push(line);
        }
      }
    }

    if (!this.isEmpty()) {
      const touchesExisting = lines.some((line) =>
        line.some((c) => this.cells.has(posKey(c.x, c.y)))
      );
      if (!touchesExisting) {
        return { ok: false, reason: "Your tiles must connect to tiles already on the board." };
      }
    }

    return { ok: true, lines };
  }

  // Given the validated `lines` from validateMove, compute the score.
  scoreMove(lines) {
    let total = 0;
    for (const line of lines) {
      const n = line.length;
      if (n < 2) continue;
      total += n;
      if (n === 6) total += 6; // Qwirkle bonus
    }
    return total;
  }

  applyMove(move) {
    for (const [x, y, t] of move) {
      this.cells.set(posKey(x, y), t);
    }
  }

  // ---------- convenience for bots/GUI ----------

  // All empty squares adjacent to at least one occupied square -- useful
  // for bots scanning for candidate placements.
  legalAnchorSquares() {
    if (this.isEmpty()) return [[0, 0]];
    const anchors = new Set();
    for (const key of this.cells.keys()) {
      const [x, y] = parseKey(key);
      for (const [dx, dy] of [[1, 0], [-1, 0], [0, 1], [0, -1]]) {
        const p = posKey(x + dx, y + dy);
        if (!this.cells.has(p)) anchors.add(p);
      }
    }
    return Array.from(anchors, parseKey);
  }

  // [minX, minY, maxX, maxY] of played tiles, or null if empty.
  bounds() {
    if (this.isEmpty()) return null;
    let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
    for (const key of this.cells.keys()) {
      const [x, y] = parseKey(key);
      if (x < minX) minX = x;
      if (y < minY) minY = y;
      if (x > maxX) maxX = x;
      if (y > maxY) maxY = y;
    }
    return [minX, minY, maxX, maxY];
  }
}
